scripts/retrieve_databundle_light.py

```diff
@@ -447,7 +448,7 @@ def _check_disabled_by_opt(config_bundle, config_enable):
 
 
 def get_best_bundles_by_category(
-    country_list, category, config_bundles, tutorial, config_enable
+    country_list, category, config_bundles, tutorial, config_enable, df
```

Comment:
The df dataframe here is not needed I believe, but rather we create it inside the function itself using the config_bundles.

I think we can reset this branch by:
git reset --hard b205e83c3b631448a778e29257eed551790b9d5c
git push --force

Be aware that after the git push, this PR will automatically close, but no problem, reopen it after you add other commits

Reply:
Hi Davide,
I have tried to create dataframe in the function itself. However, the function is called multiple times by get_best_bundles function. Therefore, I am using the data in the dataframe since the loop deletes everything in the data and creates empty one. So my solution was to create empty dataframe before entering the get_best_bundles_by_category and prevent data loss. If you have a way around with this problem I would appreciate your code help :) Also I am not I understand creating df frame using config_bundles.

Many thanks for the review. I will revise it upon your comments.

Reply:
so, I agree that the function will be called multiple times and that's no problem, the overhead is also limited.
Moreover, every dataframe refers to a different type of databundle, so that's ok.

Maybe you need a bit of contextualization; I try to add some context below. That is specified on top of the script, but also feel free to revise the docstring of that function to improve the clarity.

If you notice the databundles, they have a flag named "category".
This function that you are editing is called for every value of category to identify the sets of bundles that are needed to download by category.
Generally, therefore, it may be called four times: one for "common", one for "data", one for "cutouts" value and one for "landcover"
For every call, the dictionary dict_n_elems is filled only with those bundles that:
1. match the category value of the current execution (https://github.com/pypsa-meets-earth/pypsa-earth/blob/95a544395e44b3a159dc4267348536599430edc5/scripts/retrieve_databundle_light.py#L489)
2. match the tutorial flag: if the tutorial is true, only the tutorial bundles are checked, if else the converse (https://github.com/pypsa-meets-earth/pypsa-earth/blob/95a544395e44b3a159dc4267348536599430edc5/scripts/retrieve_databundle_light.py#L490)
3. a further check is performed to exclude bundles whose outputs are no more needed, for example if build_cutout is enabled that no more requires the cutouts flag. This is done using `_check_disabled_by_opt` https://github.com/pypsa-meets-earth/pypsa-earth/blob/95a544395e44b3a159dc4267348536599430edc5/scripts/retrieve_databundle_light.py#L491

So the first goal would be to create a df_matches that replaces the dict_n_matches and fill it with additional columns.
The comment on the following lines adds more and provides a draft structure of the revisione.

Hope this message better clarifies some doubts, if something is not clear, please, let me know :)